rework opaque type region inference

aliemjay opened the pull request:
User-facing changes are documented in [this comment](https://github.com/rust-lang/rust/pull/116891#issuecomment-1973774412).

The design document is in [this comment](https://github.com/rust-lang/rust/pull/116891#issuecomment-1836900102).

---

\- Fix Ice in check_unique; ICE -> Error; fixes #122782.
\- Ignore uncaptured lifetime args; ICE -> Pass; fixes #111906, fixes #110623, fixes #109059, fixes #122307
\- Except equal parameters from the uniqueness check; Pass -> Error; fixes #113916.
\- Check RPITs for invalid args; Pass -> Error; fixes #111935; ICE -> Error; fixes #110726.
\- Rework opaque types region inference; Pass -> Error; fixes #113971, fixes #112841.
\- Reject external lifetimes as invalid args; Pass -> Error; fixes #105498.

r? @ghost

aliemjay committed:
favor placeholders over existentials when choosing SCC representatives

... even when the existential has the least RegionVid.

universal regions (of root universe) > placeholders > existentials

The previous behavior, that chooses the minimal RegionVid index, naturally prefers universal regions over others
because they always have the least RegionVids, but there was no guranteed ordering between placeholders and existentials.

aliemjay committed:
convert all named regions in opaque types to nll vars

Do it in typeck before entering region inference routines
particularly because we will no longer be able to convert placeholders.

aliemjay committed:
rework opaque types region inference

aliemjay committed:
reject external lifetimes as invalid arguments

aliemjay commented:
@bors try

bors commented:
:hourglass: Trying commit 14ce42a539f44525e39c4e6be35d0caa4d2e8e74 with merge c679690c6334d8dd7b925210b8351178b8064f03...
<!-- homu: {"type":"TryBuildStarted","head_sha":"14ce42a539f44525e39c4e6be35d0caa4d2e8e74","merge_sha":"c679690c6334d8dd7b925210b8351178b8064f03"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6573611032/job/17859654949)
Build commit: c679690c6334d8dd7b925210b8351178b8064f03 (`c679690c6334d8dd7b925210b8351178b8064f03`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6573611032/job/17859654949"},"merge_sha":"c679690c6334d8dd7b925210b8351178b8064f03"} -->

aliemjay commented:
@craterbot check

aliemjay committed:
except equal parameters from the uniqueness check

aliemjay commented:
@bors try

bors commented:
:hourglass: Trying commit 20a516349128c45105eb6389be3b820a8e8275ec with merge e287a7de9d1b45b00b9425dcb2c5a526e3c91550...
<!-- homu: {"type":"TryBuildStarted","head_sha":"20a516349128c45105eb6389be3b820a8e8275ec","merge_sha":"e287a7de9d1b45b00b9425dcb2c5a526e3c91550"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6598476796/job/17927208883)
Build commit: e287a7de9d1b45b00b9425dcb2c5a526e3c91550 (`e287a7de9d1b45b00b9425dcb2c5a526e3c91550`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6598476796/job/17927208883"},"merge_sha":"e287a7de9d1b45b00b9425dcb2c5a526e3c91550"} -->

aliemjay commented:
I'll add some other fixes and crater them later in a single run.

@craterbot abort

aliemjay commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 350b33ee7aabbea131a63217f4f610bcde81cfe7 with merge 66abdf72fc886ae058026c1529ae9c93ebd8838a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"350b33ee7aabbea131a63217f4f610bcde81cfe7","merge_sha":"66abdf72fc886ae058026c1529ae9c93ebd8838a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6610206897/job/17954309542)
Build commit: 66abdf72fc886ae058026c1529ae9c93ebd8838a (`66abdf72fc886ae058026c1529ae9c93ebd8838a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6610206897/job/17954309542"},"merge_sha":"66abdf72fc886ae058026c1529ae9c93ebd8838a"} -->

rust-timer commented:
Queued 66abdf72fc886ae058026c1529ae9c93ebd8838a with parent aec4741d42e10673e7d5fe0f1c96d407801884e8, future [comparison URL](https://perf.rust-lang.org/compare.html?start=aec4741d42e10673e7d5fe0f1c96d407801884e8&end=66abdf72fc886ae058026c1529ae9c93ebd8838a).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

aliemjay commented:
@craterbot check

rust-timer commented:
Finished benchmarking commit (66abdf72fc886ae058026c1529ae9c93ebd8838a): [comparison URL](https://perf.rust-lang.org/compare.html?start=aec4741d42e10673e7d5fe0f1c96d407801884e8&end=66abdf72fc886ae058026c1529ae9c93ebd8838a&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=aec4741d42e10673e7d5fe0f1c96d407801884e8&end=66abdf72fc886ae058026c1529ae9c93ebd8838a&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=aec4741d42e10673e7d5fe0f1c96d407801884e8&end=66abdf72fc886ae058026c1529ae9c93ebd8838a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.9%  | [3.9%, 3.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [2.3%, 2.6%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.2% | [-4.2%, -4.2%] | 1     |
| All ❌✅ (primary)                 | 3.9%  | [3.9%, 3.9%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=aec4741d42e10673e7d5fe0f1c96d407801884e8&end=66abdf72fc886ae058026c1529ae9c93ebd8838a&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.8% | [-4.7%, -2.4%] | 5     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=aec4741d42e10673e7d5fe0f1c96d407801884e8&end=66abdf72fc886ae058026c1529ae9c93ebd8838a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 304.28 MiB -> 304.25 MiB (-0.01%)

lqd commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -66,85 +66,60 @@ impl<'tcx> RegionInferenceContext<'tcx> {
     ) -> FxIndexMap<LocalDefId, OpaqueHiddenType<'tcx>> {
```
GH doesn't allow comments on unmodified code, but I have a question about the `infer_opaque_types` function's documentation above: will this doc comment still be accurate after the changes in this PR? 

(At least it seems the "which has no `external_name` in which case we use `'empty` as the" sentence was missed when `ReEmpty` was removed last year)

I don't know this resolving of opaque type regions at all, but I have the feeling that, if that high-level description doesn't need changes when the mapping to external names has been rewritten, then it's a probable sign that the body of the function itself could use more comments explaining what it does and why (e.g. the reasons why this required changing how the SCC representatives were picked in the first place)?

aliemjay replied:
I've missed this doc comment. It indeed needs a rewrite, in addition to documenting the new changes.

oli-obk replied:
This is still unaddressed

aliemjay replied:
Oh, ReEmpty is still there. give me a sec.

lqd commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -372,38 +347,33 @@ fn check_opaque_type_well_formed<'tcx>(
-fn check_opaque_type_parameter_valid(
-    tcx: TyCtxt<'_>,
-    opaque_type_key: OpaqueTypeKey<'_>,
+fn check_opaque_type_parameter_valid<'tcx>(
```
Linking to where these opaque type parameter validity rules are documented would be useful here.

lqd commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -435,3 +405,81 @@ fn check_opaque_type_parameter_valid(
     Ok(())
 }
+
+struct LazyOpaqueTyEnv<'tcx> {
```
This likely could use a doc comment.

lqd commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -435,3 +405,81 @@ fn check_opaque_type_parameter_valid(
+        canonical_args[param1] == canonical_args[param2]
+    }
+
+    fn get_canonical_args(&self) -> ty::GenericArgsRef<'tcx> {
```
Otherwise this definitely could use a doc comment.

lqd reviewed (commented):
This isn't a draft PR, but still looks in progress, so I'm not sure whether you wanted feedback at this early stage. Apologies if it's too soon and you're still working on the things I'll leave comments about.

I personally would like to see more comments and documentation in the compiler in general, but in particular in `rustc_borrowck` here, especially on complicated areas with a low "bus factor". 

That will in turn spread knowledge around, help with reviews, and a lot more.

To be clear, I don't know the opaque type inference at all, will let the experts comment about that, while my comments will be more general (and surely less useful).

aliemjay commented:
> This isn't a draft PR, but still looks in progress, so I'm not sure whether you wanted feedback at this early stage. Apologies if it's too soon and you're still working on the things I'll leave comments about.



> I personally would like to see more comments and documentation in the compiler in general, but in particular in `rustc_borrowck` here, especially on complicated areas with a low "bus factor".

@lqd the documentation work is indeed the only blocker of this PR. I'm aiming for a dedicated rustc-dev-guide section on the subtleties of opaque types in borrowck before requesting a review or starting an FCP. And of course an early review is welcome. For now there are the commit messages and some comments in test files that might help in early review.

aliemjay commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -487,22 +489,27 @@ impl<'tcx> RegionInferenceContext<'tcx> {
-            scc_representatives[scc] = region_vid.min(prev_min);
+        let mut scc_representatives = IndexVec::from_elem_n(ty::RegionVid::MAX, num_sccs);
+
+        for (vid, def) in definitions.iter_enumerated() {
+            use NllRegionVariableOrigin as VarOrigin;
+            let scc = constraints_scc.scc(vid);
+            let repr = &mut scc_representatives[scc];
+            if *repr == ty::RegionVid::MAX {
+                *repr = vid;
+            } else if matches!(def.origin, VarOrigin::Placeholder(_))
+                && matches!(definitions[*repr].origin, VarOrigin::Existential { .. })
+            {
+                *repr = vid;
+            }
         }
```
I believe it does. We iterate over the all RegionVids *in-order*, picking the first one as a representative, which will always be the free region if there is a one in the SCC. Then we only have to care about prioritizing placeholders over existentials in order to guarantee the documented ordering.

lqd replied:
Then I'm missing where/how we now ensure that the representative is the documented minimal `RegionVid` like it was done before?

lqd replied:
@aliemjay a funny thing is that #117146 is caused by an SCC that contains a placeholder but has an existential representative — which is not very... _representative_  (🥁) — and would be fixed by this implementation :)

lqd commented:
Randomly browsing around only lead to disk space issues.

@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-116891/retry-regressed-list.txt

lqd commented:
(craterbot has returned the status to waiting-on-review while ali had set it to waiting-on-author: @rustbot author)

aliemjay commented:
** Here is a draft of what is intended to be rustc-dev-guide chapter, Please feel free to edit.**

# Opaque types region inference

In this section we discuss the various restrictions we impose on the generic arguments of opaque types when defining their hidden types
`Opaque<'a, 'b, .., A, B, ..> := SomeHiddenType`.

## Background: type arguments
For type arguments, two restrictions are necessary: each type argument must be (1) a type parameter and (2) is unique among the generic arguments.
Example of case (1):
```rust
type Opaque<X> = impl Sized;

// `T` is a type paramter.
// Opaque<T> := ();
fn good<T>() -> Opaque<T> {}

// `()` is not a type parameter.
// Opaque<()> := ();
fn bad() -> Opaque<()> {} //~ ERROR
```

Example of case (2):
```rust
type Opaque<X, Y> = impl Sized;

// `T` and `U` are unique in the generic args.
// Opaque<T, U> := T;
fn good<T>(t: T) -> Opaque<T, U> { t }

// `T` appears twice in the generic args.
// Opaque<T, T> := T;
fn bad<T>(t: T) -> Opaque<T, T> { t } //~ ERROR
```
**Motivation:** In the first case `Opaque<()> := ()`, the hidden type is ambiguous because it is compatible with two different interpretaions: `Opaque<X> := X` and `Opaque<X> := ()`. Similarily for the second case `Opaque<T, T> := T`, it is ambiguous whether it should be interpreted as `Opaque<X, Y> := X` or as `Opaque<X, Y> := Y`. Because of this ambiguity, both cases are rejected as invalid defining use.

## Uniqueness restriction

> With the exception below, all lifetime args must be unique in the arguments list and must not be `'static`. This is in order to avoid an ambiguity with hidden type inference similar to the case of type parameters. For example, the invalid defining use below `Opaque<'static> := Inv<'static>` is compatible with both `Opaque<'x> := Inv<'static>` and `Opaque<'x> := Inv<'x>`.
```rust
type Opaque<'x> = impl Sized + 'x;
type Inv<'a> = Option<*mut &'a ()>;

fn good<'a>() -> Opaque<'a> { Inv::<'static>::None }

fn bad() -> Opaque<'static> { Inv::<'static>::None }
//~^ ERROR

// This is also invalid because `'a` is *semantically* equal to `'static`.
fn still_bad<'a: 'static>() -> Opaque<'a> { Inv::<'static>::None }
//~^ Should error!
```

```rust
type Opaque<'x, 'y> = impl Trait<'a, 'b>;

fn good<'a, 'b>() -> Opaque<'a, 'b> {}

fn bad<'a>() -> Opaque<'a, 'a> {}
//~^ ERROR

// This is also invalid because `'a` and `'b` are *semantically* equal.
fn still_bad<'a: 'b, 'b: 'a>() -> Opaque<'a, 'b> {}
//~^ Should error!
```

**Status:** 
- [ ] For TAIT, semantic equality of lifetimes are not always used in this check. This leads to the surprising behavior in [#112841][].
- [ ] For RPIT, this check is not implemented at all (see next section for why).

[#112841]: https://github.com/rust-lang/rust/issues/112841

## A proposed exception to uniqueness rule

- As an exception to the uniqueness rule, above is when the bounds on opaque type's definition require that a lifetime parameter be equal to another one or to the `'static` lifetime.
```rust
// The definition requires `'x` to be equal to `'static`.
type Opaque<'x: 'static> = impl Sized + 'x;

fn good() -> Opaque<'static> {}
```

**Motivation:** an attempt to implement the uniqueness restriction for RPITs resulted in a [breakage found via crater]( https://github.com/rust-lang/rust/pull/112842#issuecomment-1610057887). Thi can be mitigated by this e to the rule. An example of the the code that would otherwise break:
```rust
struct Type<'a>(&'a ());
impl<'a> Type<'a> {
    // `'b == 'a`
    fn do_stuff<'b: 'a>(&'b self) -> impl Trait<'a, 'b> {}
}
```

Why is this correct: 

**Status:** 
- [ ] Not implemented. Issue #113916.

## Universal lifetimes restriction

- Only universally quantified lifetimes are allowed in the opaque type arguments. This includes lifetime parameters and placeholders.

```rust
type Opaque<'x> = impl Sized + 'x;

fn test() -> Opaque<'a> {
    // `Opaque<'empty> := ()`
    let _: Opaque<'_> = ();
    //~^ ERROR
}
```

**Motivation:** This makes the lifetime and type argums behave consistently but this is only as a bonus. The real reason behind this restriction is purely technical, as the "member constraints" algorithm faces a fundamental limitation: When encountering an opaque type definition `Opaque<'?1> := &'?2 u8`, a member constraint `'?2 :belongs_to: ['static, '?1]` is registered. In order for the algorithm to pick the right choice, the *complete* set of "outlives" relationships between the choice regions `['static, '?1]` must already be known *before* doing the region inference. This can be satisfied only if each choice region is either:
1. a universal region, i.e. `RegionKind::Re{EarlyParam,LateParam,Placeholder,Static}`, because the relations between universal regions are completely known, prior to region inference, from the explicit and implied bounds.
1. or an existential region that is "strictly equal" to a universal region.  Strict lifetime equality is defined below and is required here because it is the only type of equality that can be evaluated prior to full region inference.

**Strict lifetime equality:** We say that two lifetimes are strictly equal if there are bidirectional outlives constraints between them. In NLL terms we say that the lifetimes are part of the same [SCC]. Importantly this type of equality can be evaluated prior to full region inference (but of course after constraint collection). The other type of equality is when region inference ends up giving two lifetimes variables the same value even if they are not strictly equal. See [#113971] for how we currently mess this up.

**Status:**
- [ ] Strict lifetime equality is not yet enforced and this gives rise to arguably behavior (see [#113971][]). 
- [ ] For RPIT, if the lifetime is not universal the compiler panics with message `opaque type with non-universal region substs` [#110726](https://github.com/rust-lang/rust/issues/110726)

[#113971]: https://github.com/rust-lang/rust/issues/113971

**interaction with "once modulo regions" restriction** In the example above, note the opaque type in the signature is `Opaque<'a>` and the one in the invalid defining use is `Opaque<'empty>`. In the proposed MiniTAIT plan, namely the ["once modulo regions"][#116935] rule, we already disallow this.
Although it might appear that "universal lifetimes" restriction becomes redundant as it logically follows from "MiniTAIT" restrictions, the subsequent related discussion on lifetime equality and closures remains relevant.

[#116935]: https://github.com/rust-lang/rust/pull/116935


## Closure restrictions

- When the opaque type is defined in a closure/coroutine/inline-const body, universal lifetimes that are "external" to the closure are not allowed in the opaque type arguments. External regions are defined in [`RegionClassification::External`](https://github.com/rust-lang/rust/blob/caf730043232affb6b10d1393895998cb4968520/compiler/rustc_borrowck/src/universal_regions.rs#L201).

Example: (This one happens to compile in the current nightly but more practical examples are already rejected with confusing errors. See [#105498][] for more).
```rust
type Opaque<'x> = impl Sized + 'x;

fn test<'a>() -> Opaque<'a> {
    let _ = || {
        // `'a` is external to the closure
        let _: Opaque<'a> = ();
        //~^ Should be an error!
    };
    ()
}
```

**Motivation:** In closure bodies, external lifetimes, although being categorized as "universal" lifetimes, behave more like existential lifetimes in that the relations between them are not known ahead of time, instead their values are inferred just like existential lifetimes and the requirements are propagated back to the parent fn. This breaks the member constraints algorithm as described above:
> In order for [the algorithm] to pick the right choice, the complete set of “outlives” relationships between the choice regions ['static, '?1] must already be known before doing the region inference

Here is an example that details how 

```rust
type Opaque<'x, 'y> = impl Sized;

// 
fn test<'a, 'b>(s: &'a str) -> impl FnOnce() -> Opaque<'a, 'b> {
    move || { s }
    //~^ ERROR hidden type for `Opaque<'_, '_>` captures lifetime that does not appear in bounds
}

// The above closure body is desugared into something like:
fn test::{closure#0}(_upvar: &'?8 str) -> Opaque<'?6, '?7> {
    return _upvar
}

// where `['?8, '?6, ?7] are universal lifetimes *external* to the closure.
// There are no known relations between them *inside* the closure.
// But in the parent fn it is known that `'?6: '?8`.
//
// When encountering an opaque definition `Opaque<'?6, '?7> := &'8 str`,
// The member constraints algotithm does not know enough to safely make `?8 = '?6`.
// For this reason, it errors with a sensible message:
// "hidden type captures lifetime that does not appear in bounds".
```

Without this restrictions error messages are consfusing and, more impotantly, there is a risk that we accept code the we would likely break in the future because member constraints are super broken in closures.

**Status:**
- [ ] Unimplemented, leading to surprising behavior and risking forward compatibility. [#105498][]

[#105498]: https://github.com/rust-lang/rust/issues/105498


**Output types:** I believe the most common scenario where this causes issues in real-world code is with closure/async-block output types. It is worth noting that there is a discrepancy betweeen closures and async blocks that further demonstrates this issue and is attributed to the [hack of `replace_opaque_types_with_inference_vars`](https://github.com/rust-lang/rust/blob/9cf18e98f82d85fa41141391d54485b8747da46f/compiler/rustc_hir_typeck/src/closure.rs#L743), which is applied to futures only.
```rust
type Opaque<'x> = impl Sized + 'x;
fn test<'a>() -> impl FnOnce() -> Opaque<'a> {
    // Output type of the closure is Opaque<'a>
    // -> hidden type definition happens *inside* the closure
    // -> rejected.
    move || {}
    //~^ ERROR expected generic lifetime parameter, found `'_`
}
```
```rust
use std::future::Future;
type Opaque<'x> = impl Sized + 'x;
fn test<'a>() -> impl Future<Output = Opaque<'a>> {
    // Output type of the async block is unit `()`
    // -> hidden type definition happens in the parent fn
    // -> accepted.
    async move {}
}
```

aliemjay commented:
r? @compiler-errors @oli-obk (owners of MinTAIT) for early input about the design. It eventually needs FCP.

cc @lcnr might also be interested.

oli-obk commented:
I won't get to this before January, but if @compiler-errors considers it ready before then, don't wait on my approval

fmease commented:
> I was wondering what == means in the PR description?

“Same as line above”. https://en.wikipedia.org/wiki/Ditto_mark

compiler-errors commented:
r? oli-obk

aliemjay commented:
This FCP aims to get the types team approval of the revised rules regarding opaque types region inference. The details are documented in [this comment]() and they were discussed in a team meeting at https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/2024-01-22.20deep.20meeting.

In addition to documenting the status quo, the new rules have few changes:
1. "Uniqueness restriction" now applies to RPIT as well - previously enforced only for TAIT. *Breaking change*. Fixes #111935. 
1. "A proposed exception to uniqueness rule". Strictly accepts more code. Helps to mitigate the impact of the the former change. Fixes #113916.
1. "Strict lifetime equality". *Breaking change*. Fixes #113971.
1. "Closure restrictions". *Breaking change*. Fixes #105498.

The [crater run](https://github.com/rust-lang/rust/pull/116891#issuecomment-1787634975) showed no regressions for this change. It is possible to hit a regression using only the stable RPIT, but I think it is very unlikely to appear in real-world code as shown by the crater result. Unstable code using TAIT is the most affected.

@rfcbot merge

BoxyUwU commented:
For the `A proposed exception to uniqueness rule` section I think it'd be good to explicitly call out that `&'b self` introduces a `Type<'a>: 'b` and therefore `'a: 'b` bound which allows `'b == 'a` to hold. I didn't notice the implied bound at first which made that section rather confusing ^^

lcnr commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -80,8 +80,16 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                             .universal_regions()
-                            .filter(|&ur| self.universal_region_relations.equal(vid, ur))
-                            .find_map(|ur| self.definitions[ur].external_name),
+                            .filter(|&ur| {
```
please add a comment here explaining why these `filter` exist

lcnr reviewed (commented):
while I would still like to see some additional comments in some places, this PR is absolutely wonderful. Thank you for your good work :heart:

BoxyUwU commented:
this was a really like nice fcp comment btw

oli-obk commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -413,3 +405,81 @@ fn check_opaque_type_parameter_valid<'tcx>(
+struct LazyOpaqueTyEnv<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    canonical_args: std::cell::Cell<Option<ty::GenericArgsRef<'tcx>>>,
```
use OnceCell

lcnr commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -413,3 +405,81 @@ fn check_opaque_type_parameter_valid<'tcx>(
+
+        let &Self { tcx, def_id, .. } = self;
+        let origin = tcx.opaque_type_origin(def_id);
+        let defining_use_anchor = match origin {
+            hir::OpaqueTyOrigin::FnReturn(did) | hir::OpaqueTyOrigin::AsyncFn(did) => did,
+            hir::OpaqueTyOrigin::TyAlias { .. } => tcx.impl_trait_parent(def_id),
+        };
+        let param_env = tcx.param_env(defining_use_anchor);
```
i think we may want to go the other way and ICE when using the `param_env` for an unexpected `DefKind`

oli-obk replied:
I gave this a try... the lint architecture calls param env on all items, including hir::OpaqueDef during recursive visiting.

We could check it outside the query, but then we're checking it inside for every param_env call, as well as inside for the assert.

lcnr replied:
:thinking: that is non-trivial, could be potentially fixable by changing the lint architecture to use the correct "root env" or whatever :shrug:

aliemjay committed:
ignore uncaptured lifetimes when checking opaques

aliemjay committed:
check RPITs for invalid args

aliemjay committed:
ignore error params

aliemjay committed:
fix ICE in check_unique

aliemjay committed:
simplify check_unique

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8432060953/job/23090496846) [(plain)](https://github.com/rust-lang/rust/commit/020ed558b656c12589d75c5a5ed2b2e0763243cd/checks/23090496846/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.5s done
#16 DONE 34.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
normalized stderr:
error[E0792]: expected generic lifetime parameter, found `'static`
##[error]  --> $DIR/defining-use-captured-non-universal-region.rs:10:18
   |
LL | fn foo<'a>() -> impl Sized + 'a {
   |                              -- cannot use static lifetime; use a bound lifetime instead or remove the lifetime parameter from the opaque type
LL |     #[cfg(statik)]
LL |     let i: i32 = foo::<'static>();

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0792`.
For more information about this error, try `rustc --explain E0792`.



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.statik/defining-use-captured-non-universal-region.statik.stderr
To only update this specific test, also pass `--test-args impl-trait/defining-use-captured-non-universal-region.rs`


error in revision `statik`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/impl-trait/defining-use-captured-non-universal-region.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "statik" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.statik" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.statik/auxiliary"
--- stderr -------------------------------
error[E0792]: expected generic lifetime parameter, found `'static`
##[error]  --> /checkout/tests/ui/impl-trait/defining-use-captured-non-universal-region.rs:10:18
   |
   |
LL | fn foo<'a>() -> impl Sized + 'a {
   |                              -- cannot use static lifetime; use a bound lifetime instead or remove the lifetime parameter from the opaque type
LL |     #[cfg(statik)]
LL |     let i: i32 = foo::<'static>();

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0792`.
```

</details>

aliemjay commented:
Rebased, refactored and reordered commits to be more independent.

New commits: "fix ICE in check_unique" and "Simplify check_unique"

Doc comments were added to the commits "except equal params from uniqueness check" and "rework opaque types region inference".

@rustbot ready

compiler-errors commented:
This is awesome work! I want to give perhaps @oli-obk an opportunity to read this, though.

compiler-errors commented on tests/ui/type-alias-impl-trait/equal-lifetime-params-ok.rs:
```diff
@@ -0,0 +1,50 @@
+// FIXME: description
```
FIXME

compiler-errors commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -18,99 +17,45 @@ use rustc_trait_selection::traits::ObligationCtxt;
+    /// their `external_name` giving `_Return<'a>`. This step is a bit involved.
+    /// See the [rustc-dev-guide chapter] for more info.
+    ///
+    /// [rustc-dev-guide chapter]: FIXME(aliemjay)
```
Do you plan on uploading this?

compiler-errors commented on compiler/rustc_borrowck/src/region_infer/opaque_types.rs:
```diff
@@ -422,42 +369,45 @@ fn check_opaque_type_well_formed<'tcx>(
-    opaque_type_key: OpaqueTypeKey<'_>,
+/// Opaque type parameter validity rules are documented in a [rustc-dev-guide chapter].
+///
+/// [rustc-dev-guide chapter]: FIXME(aliemjay)
```
This too

oli-obk commented:
lgtm (though https://github.com/rust-lang/rust/pull/116891#discussion_r1375253753 is still unaddressed)

aliemjay commented:
> lgtm (though [#116891 (comment)](https://github.com/rust-lang/rust/pull/116891#discussion_r1375253753) is still unaddressed)

I think I have amended the commit "rework opaque types region inference" the address it.

oli-obk reviewed (approved):
r=me,errs with dev-guide urls updated

aliemjay committed:
fixup except equal params from

aliemjay committed:
remove test FIXME re once-module-region

aliemjay commented:
@bors r=compiler-errors,oli-obk rollup=never

@rustbot label +relnotes

bors commented:
:pushpin: Commit 4008a39e14699cd2a4d3da386ac4b1f84574eee3 has been approved by `compiler-errors,oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors,oli-obk 4008a39e14699cd2a4d3da386ac4b1f84574eee3 -->
<!-- homu: {"type":"Approved","sha":"4008a39e14699cd2a4d3da386ac4b1f84574eee3","approver":"compiler-errors,oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8463264457/job/23185907306) [(plain)](https://github.com/rust-lang/rust/commit/4008a39e14699cd2a4d3da386ac4b1f84574eee3/checks/23185907306/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.1s done
#17 DONE 34.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
1 error[E0792]: expected generic lifetime parameter, found `'_`
-   --> $DIR/defining-use-captured-non-universal-region.rs:14:18
+   --> $DIR/defining-use-captured-non-universal-region.rs:13:18
3    |
4 LL | fn foo<'a>() -> impl Sized + 'a {
5    |                              -- this generic parameter must be used with a generic lifetime parameter

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.infer/defining-use-captured-non-universal-region.infer.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args impl-trait/defining-use-captured-non-universal-region.rs`

error in revision `infer`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/impl-trait/defining-use-captured-non-universal-region.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "infer" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.infer" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.infer/auxiliary"
--- stderr -------------------------------
error[E0792]: expected generic lifetime parameter, found `'_`
##[error]  --> /checkout/tests/ui/impl-trait/defining-use-captured-non-universal-region.rs:13:18
   |
   |
Build completed unsuccessfully in 0:14:04
LL | fn foo<'a>() -> impl Sized + 'a {
   |                              -- this generic parameter must be used with a generic lifetime parameter
...
LL |     let i: i32 = foo::<'_>();

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0792`.
---
1 error[E0792]: expected generic lifetime parameter, found `'static`
-   --> $DIR/defining-use-captured-non-universal-region.rs:10:18
+   --> $DIR/defining-use-captured-non-universal-region.rs:9:18
3    |
4 LL | fn foo<'a>() -> impl Sized + 'a {
5    |                              -- cannot use static lifetime; use a bound lifetime instead or remove the lifetime parameter from the opaque type

The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.statik/defining-use-captured-non-universal-region.statik.stderr
To only update this specific test, also pass `--test-args impl-trait/defining-use-captured-non-universal-region.rs`


error in revision `statik`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/impl-trait/defining-use-captured-non-universal-region.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "statik" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.statik" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/defining-use-captured-non-universal-region.statik/auxiliary"
--- stderr -------------------------------
error[E0792]: expected generic lifetime parameter, found `'static`
##[error]  --> /checkout/tests/ui/impl-trait/defining-use-captured-non-universal-region.rs:9:18
   |
   |
LL | fn foo<'a>() -> impl Sized + 'a {
   |                              -- cannot use static lifetime; use a bound lifetime instead or remove the lifetime parameter from the opaque type
LL |     #[cfg(statik)]
LL |     let i: i32 = foo::<'static>();

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0792`.
---
1 error[E0792]: expected generic lifetime parameter, found `'static`
-   --> $DIR/non-defining-use-lifetimes.rs:14:16
+   --> $DIR/non-defining-use-lifetimes.rs:13:16
3    |
4 LL |     fn foo<'a>(_: Lt<'a>) -> impl Sized + 'a {
5    |                                           -- cannot use static lifetime; use a bound lifetime instead or remove the lifetime parameter from the opaque type
7    |                ^^
8 
9 error[E0792]: expected generic lifetime parameter, found `'_`
-   --> $DIR/non-defining-use-lifetimes.rs:23:16
-   --> $DIR/non-defining-use-lifetimes.rs:23:16
+   --> $DIR/non-defining-use-lifetimes.rs:22:16
11    |
12 LL |     fn foo<'a>(_: Lt<'a>) -> impl Sized + 'a {
13    |                                           -- this generic parameter must be used with a generic lifetime parameter
15    |                ^^
16 
17 error: non-defining opaque type use in defining scope
-   --> $DIR/non-defining-use-lifetimes.rs:33:16
-   --> $DIR/non-defining-use-lifetimes.rs:33:16
+   --> $DIR/non-defining-use-lifetimes.rs:32:16
19    |
20 LL |         let _: () = foo(Lt::<'a>::None, Lt::<'a>::None);

22    |
23 note: lifetime used multiple times
-   --> $DIR/non-defining-use-lifetimes.rs:32:58
-   --> $DIR/non-defining-use-lifetimes.rs:32:58
+   --> $DIR/non-defining-use-lifetimes.rs:31:58
25    |
26 LL |     fn foo<'a, 'b>(_: Lt<'a>, _: Lt<'b>) -> impl Sized + 'a + 'b {
27    |                                                          ^^   ^^

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/impl-trait/rpit/non-defining-use-lifetimes/non-defining-use-lifetimes.stderr
To update references, rerun the tests and pass the `--bless` flag
---
--- stderr -------------------------------
error[E0792]: expected generic lifetime parameter, found `'static`
##[error]  --> /checkout/tests/ui/impl-trait/rpit/non-defining-use-lifetimes.rs:13:16
   |
LL |     fn foo<'a>(_: Lt<'a>) -> impl Sized + 'a {
   |                                           -- cannot use static lifetime; use a bound lifetime instead or remove the lifetime parameter from the opaque type
LL |         let _: () = foo(Lt::<'static>::None);

error[E0792]: expected generic lifetime parameter, found `'_`
##[error]  --> /checkout/tests/ui/impl-trait/rpit/non-defining-use-lifetimes.rs:22:16
   |
   |
LL |     fn foo<'a>(_: Lt<'a>) -> impl Sized + 'a {
   |                                           -- this generic parameter must be used with a generic lifetime parameter
LL |         let _: () = foo(Lt::<'_>::None);

error: non-defining opaque type use in defining scope
##[error]  --> /checkout/tests/ui/impl-trait/rpit/non-defining-use-lifetimes.rs:32:16
   |
   |
LL |         let _: () = foo(Lt::<'a>::None, Lt::<'a>::None);
   |
note: lifetime used multiple times
  --> /checkout/tests/ui/impl-trait/rpit/non-defining-use-lifetimes.rs:31:58
   |
   |
LL |     fn foo<'a, 'b>(_: Lt<'a>, _: Lt<'b>) -> impl Sized + 'a + 'b {
   |                                                          ^^   ^^
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0792`.
------------------------------------------
```

</details>

aliemjay commented:
@bors r-

Waiting on CI.

aliemjay commented:
@bors r=compiler-errors,oli-obk rollup=never

bors commented:
:pushpin: Commit 59c217fed2bc74f3685bf3461674f6b3c7e113c3 has been approved by `compiler-errors,oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors,oli-obk 59c217fed2bc74f3685bf3461674f6b3c7e113c3 -->
<!-- homu: {"type":"Approved","sha":"59c217fed2bc74f3685bf3461674f6b3c7e113c3","approver":"compiler-errors,oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 59c217fed2bc74f3685bf3461674f6b3c7e113c3 with merge 551abd65bee759c1fd1cc0acd5ba7e5723823c54...
<!-- homu: {"type":"BuildStarted","head_sha":"59c217fed2bc74f3685bf3461674f6b3c7e113c3","merge_sha":"551abd65bee759c1fd1cc0acd5ba7e5723823c54"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8465436585/job/23196523031)
Approved by: compiler-errors,oli-obk
Pushing 551abd65bee759c1fd1cc0acd5ba7e5723823c54 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors,oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8465436585/job/23196523031"},"merge_sha":"551abd65bee759c1fd1cc0acd5ba7e5723823c54"} -->

rust-timer commented:
Finished benchmarking commit (551abd65bee759c1fd1cc0acd5ba7e5723823c54): [comparison URL](https://perf.rust-lang.org/compare.html?start=d0e8cbbd99ccc1be12b1aa743521ed89268e59a2&end=551abd65bee759c1fd1cc0acd5ba7e5723823c54&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d0e8cbbd99ccc1be12b1aa743521ed89268e59a2&end=551abd65bee759c1fd1cc0acd5ba7e5723823c54&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d0e8cbbd99ccc1be12b1aa743521ed89268e59a2&end=551abd65bee759c1fd1cc0acd5ba7e5723823c54&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.5%, -0.5%] | 10    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.1% | [-1.5%, -0.5%] | 10    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d0e8cbbd99ccc1be12b1aa743521ed89268e59a2&end=551abd65bee759c1fd1cc0acd5ba7e5723823c54&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.4%, 0.8%] | 3     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.4%, 0.8%] | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d0e8cbbd99ccc1be12b1aa743521ed89268e59a2&end=551abd65bee759c1fd1cc0acd5ba7e5723823c54&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 669.796s -> 670.635s (0.13%)
**Artifact size**: 315.69 MiB -> 315.72 MiB (0.01%)
